### PR TITLE
Add yarn install to the bin/setup script

### PR DESCRIPTION
**Pull request by ZePedroResende**

### CODE REVIEW

File: bin/setup [hunk 1]
```diff
@@ -39,9 +39,8 @@ else
 fi
 
 if not_installed "yarn"; then
-  pp_error "error" "yarn was not detected on this system, you need it to install frontend dependencies"
-
-  exit -1
+  pp_info "installling yarn"
+  npm install -g yarn
```
Comment (jfranciscosousa) [inline bin/setup:43]: Wait you were talking about this yarn? I don't this is a good idea.

First, the `yarn` team does not recommend people installing it via `npm`.
Second, each user should install their tools as they see fit. There are people running on `yarn 2` and using `yarn` version managers. I don't think we should meddle with this at all.

Comment (DavideSilva) [reply bin/setup:43]: Yup. I agree with @jfranciscosousa.

Some old project was doing this and it messed up my system, specially with `asdf` versions of Node

Comment (ZePedroResende) [reply bin/setup:43]: We have a opinionated starter, with that reasoning we shouldn't use asdf also, we have a self contained version of node that is locally linked there is no way for things to get messed up with this. We can change the way yarn is installed to the recommend way , but what you said about people chosing what too use doesn't seem to have a place on a opinionated template with local versions of node that the whole reason to exist is to sandbox the system version and dependencies, we should use this too streamline the install process

Comment (jfranciscosousa) [reply bin/setup:43]: We use `asdf` here so you cannot install `yarn` that way. It's that simple. It will mess up paths **especially** if you are using ASDF. I don't **like** having global packages with yarn given that each version of node has its own packages so it's pretty much useless. In fact, I only use yarn because it's just faster than NPM. People are migrating back to npm just because of the `yarn 2` fiasco.

If someone has `asdf`, is pretty much a given that they will use `asdf` for node and elixir. But we are NOT elixir or node in any other way, neither are we forcing them to use `asdf`. Here you force people to use `npm` to install yarn. 

These are the kinds of things that should reside in a binary. And for binaries, you do it yourself. The starter was never about tooling, it's about code and config. The mandatory setup and server scripts are nice to have, and not everyone uses them.

My final take is that we can tell people to install the binary and link it https://classic.yarnpkg.com/en/docs/install/#debian-stable

Or we can tell people to install via the non recommended way (for classic yarn), with `npm`.

Comment (jfranciscosousa) [reply bin/setup:43]: Copy pasting a suggested command by us is not the end of the world, but at least you give them a choice.

About forcing `asdf`... outside of Subvisual I don't know a single person that uses this. They just use Docker or install themselves. They either match the versions on the repo or not at all. Any single person that has `asdf` will always run the install. The rest won't even care.

Comment (ZePedroResende) [reply bin/setup:43]:
```
We use asdf here so you cannot install yarn that way. It's that simple.
```
What do you mean , if you use asdf and do a global you install it to the bin folder for that specific version of node inside asdf , there is no problem there. You dont in any way mess up paths anywhere.

Comment (ZePedroResende) [reply bin/setup:43]:
```
If someone has asdf, is pretty much a given that they will use asdf for node and elixir. But we are NOT installing elixir or node in any other way, neither are we forcing them to use asdf. Here you force people to use npm to install yarn.
```

This statement just doesn't make sense, how do you use asdf and npm on the same script but are forcing one and not the other. Npm comes in the asdf install, its bundle the users doesn't need to install anything more, npm install -g will install a global to this sandboxed version , there is not problem in the path , its a specific thing for project. If we are opinionated we should be opinionated in everything end of question .

Comment (ZePedroResende) [reply bin/setup:43]: However who i said before i'm ok with using the recommended install script from the yarn team , however i believe that it should be installed because of the reasons stated before

Comment (DavideSilva) [reply bin/setup:43]: The `bin/setup` script is simply a helper to setup a project. Not a script to setup the dev environment. There might someone who doesn't want to run the script and prefers to setup the project manually. As there might be someone who already has their environment setup the way they want and do not want to install `yarn`. The `bin/setup` script follows our recommended way of setting up a project, but it is just that, a recommendation.

Also, we are not forcing anyone to use `asdf`. If you want to use the script, it's true that you need to install it, as well as `yarn`, but it is not mandatory. You can use what you prefer.

> If we are opinionated we should be opinionated in everything end of question
But this script is the exact proof that we are opinionated. We use `asdf` for version management and `yarn` for dependencies. What we are not is authoritarian and force everyone to use what we use just because.

Comment (jfranciscosousa) [reply bin/setup:43]: > What do you mean , if you use asdf and do a global you install it to the bin folder for that specific version of node inside asdf , there is no problem there. You dont in any way mess up paths anywhere.

Tell that to Davide. And still, you install `yarn` for a single version of `node`. Is that really desirable? You change node version and you'll have to install `yarn` all over again. If this is really opinionated then the opinion that is better is: install the binary with this **link**.

> This statement just doesn't make sense, how do you use asdf and npm on the same script but are forcing one and not the other. Npm comes in the asdf install, its bundle the users doesn't need to install anything more, npm install -g will install a global to this sandboxed version , there is not problem in the path , its a specific thing for project. If we are opinionated we should be opinionated in everything end of question .

In fact, it does make sense. I am not theorizing here. These things happened to me and other people in the ecosystem.

It's not about forcing `npm`, it's about enforcing using a package manager `npm` to install another package manager `yarn` that would be better off installed as a binary. I've debugged way too many problems with other people (Davide, Filipe, Tânia, Cracel), because of installing `yarn` via `npm` + `asdf`. 

With that then you should install `postgres` also. 

And by the way, In no place, we say the template is opinionated. It's just a thiiiiin layer of stuff over a regular `phoenix` project. And this has been used before by other people, who prefer to install `yarn` themselves, and even `node` and `elixir` themselves (Zamith/Pedro are two examples). This is not a streamlined do it all script. It's just a helping tool. You don't want to run with our tools, then just do it yourself. In your case installing `yarn` is as simple as just copy pasting a link, and be done with it. Install this per `node` version via `asdf` in my opinion is asking for trouble, trouble that actually happened in the past, we are not theorizing here. **All of the above happened to multiple people in the past, including me**.

It's just the way these scripts were written before I got here in the first place. We expect `bundler` and `rbenv` to be installed in a `ruby` environment. In the React world, I just expect people to install and use `yarn` nowadays.

And no, we will not run the install script by the `yarn` team. We tell people to do it themselves. If `asdf` is not present, we warn folks of it and let the script crash when it should, by either triggering `mix` or `yarn`. If `yarn` is not present we stop and tell people about it.

My last take:
- Link to install instructions, yarn binary
- Also present the `npm install -g yarn` with a disclaimer, just saying that we prefer the first.

Comment (ZePedroResende) [reply bin/setup:43]: @DavideSilva is the script for our setup so there is no problem to install yarn , we are not forcing anyone here to run it. If you run our script you simply need it because you just installed a new version of node it should be automatic in a opinionated script instead of running it once , installing something manually that could be done automatically because every one is gonna do it and run the script again

Comment (ZePedroResende) [reply bin/setup:43]: @jfranciscosousa 

> Tell that to Davide. And still, you install yarn for a single version of node. Is that really desirable? You change node version and you'll have to install yarn all over again. If this is really opinionated then the opinion that is better is: install the binary with this link.

I personally prefer it this way , every thing is separated and sandboxed from the global path and system version. However i like we already defined this script is just for the people that want to run it its not required so having yarn install even by the generic curl | bash that the yarn team provides shouldn't be a problem. 

> . This is not a streamlined do it all script. It's just a helping tool. You don't want to run with our tools, then just do it yourself

So we can add yarn because you're gonna do it because you just installed a local node version and you don't have it linked, again don't this just make me more certain that we need the install yarn here . If you dont like it do it yourself by hand else its automated even.

TL DR;
* There is no reason at all to not add the install yarn here 
* I'm completely fine adding the generic curl| bash
* There is no sense to add a manual command that you will need using the script when we are using the philosophy if you want it differently do it by hand 
* If you use the script you need yarn and we can give a sensibly way of installing yarn by default.

Comment (jfranciscosousa) [reply bin/setup:43]: I really think it's not cool to change something given your, singular, personal experience. Especially when I said, that in the past when my scripts would install `yarn` for you (like this), people would end up with broken global packages down the line.

And again. The point of having these repos has nothing to do with "automating" installs. It's just code.

Linking to their page is better because install instructions might change in the future (on their side), but if you really insist on automating, just add their bash install script.

Comment (DavideSilva) [reply bin/setup:43]: @ZePedroResende yes, there is a problem in adding the installation of Yarn to the script. And you just gave a perfect example.

You prefer to have each version of Node have its own sandboxed Yarn installation. That's fine, you choose how you want your environment. However, I installed Yarn via Homebrew. If I were to use the script with Yarn being installed in it, that would mess my system, as it happened in the past. Exactly the same. 

I ran a `bin/setup` of an old project, which installed Yarn via the `npm install -g yarn`. That caused my local env to start complaining I had no Yarn installed when I switched Node versions between projects.
It caused the executable for Yarn to only look for shim versions of it, in the `asdf` folder.
And it didn't find any because I don't install Yarn for each and every one of them.

This happened to me, not once but twice. I had a perfectly working env that started to malfunction simply because the `bin/setup` script as doing more than it should.

> I personally prefer it this way

Also, you touched a very important point here. You prefer it this way. I personally don't. What we should not do is force people either way.

I'm all for automation but doing something in a script that might break other people's environment is crossing a line that I'm not comfortable with.

Comment (DavideSilva) [reply bin/setup:43]: Also, let's imagine these two scenarios:

1. We don't install Yarn
2. We install Yarn

We each start with a fresh environment, where we both don't have the required Node version for this project. Only difference is that I have Yarn installed via Homebrew and you don't. We both run the `bin/setup`.

Scenario 1:
- For me, `asdf` installs the required Node version and since it detects I have Yarn, it starts installing the dependencies.
- For you, `asdf` installs the required Node version and since you don't have Yarn, the scripts stops and tells you to install Yarn before proceeding. You install Yarn via `npm install -g yarn`, rerun the `bin/setup` script and it installs the dependencies, now that detects you have Yarn installed.

Scenario 2:
- For me, `asdf installs the required Node version, installs Yarn via NPM, breaks my Homebrew installation of Yarn, installs the rest of the dependencies.
- For you, `asdf installs the required Node version, installs Yarn via NPM and installs the rest of the dependencies.

Scenario 1 for me is the optimal path, for you it forces to run a command and rerun the script.
Scenario 2 for you is the optimal path but breaks the environment for me.

This is assuming a clean environment. If you already happen to have Yarn installed for that Node version, the script would behave as expected.

Comment (jfranciscosousa) [reply bin/setup:43]: Scenario 2 doesn't happen. The script checks for an existing yarn `installation`. Only if I don't have yarn (forgot to install it somehow)

Comment (DavideSilva) [reply bin/setup:43]: Right. I was thinking of that old project that simply installed Yarn from `npm install` without any sort of check. 
Still, this is dangerous and the point still stands. 

We ask the user to install `asdf` if it is not found, we should ask the same for `yarn`.

@ZePedroResende have you [tried this](https://github.com/asdf-vm/asdf-nodejs#default-npm-packages)? This way, every time you install a new Node version, it can automatically install Yarn for you. No need to change the script

Comment (ZePedroResende) [reply bin/setup:43]: @DavideSilva @DavideSilva i think all the problems you are presenting are simply masseuse of the global path and incorrect usage of system versions of node (because yes homebrew when installing yarn installs a global version of node and there's where your problem comes from @DavideSilva , its a docummented problem to solve this do `brew install yarn --without-node`). There is absolutly no reason to not use sandboxed versions of yarn in my opinion, but in the end of the day that is all my opinion. 
Because i'm in the minority here im closing this.

Comment (DavideSilva) [reply bin/setup:43]: Leaving this here just so that it is documented.

No, `brew install yarn --without-node` is not a documented solution to the problem because [it no longer works](https://github.com/yarnpkg/website/issues/913). Might have been once but not anymore.

Installing Yarn via NPM [is not recommend](https://classic.yarnpkg.com/en/docs/install#alternatives-stable). The Yarn team stated this themselves as a potential source of security problems. They recommend the "installation method best suited to your operating system" instead.

About misusing the global path. In this instance I believe this is not the case. I don't remember exactly where I saw it (will link it here if I end up finding it) but I am 99% sure that Yarn installed via brew doesn't interfere with Node and works independently of it.

Comment (jfranciscosousa) [reply bin/setup:43]: When installing `yarn` with `npm` via `asdf` you put the `yarn` symlink on the shims folder of `asdf`. Every time you switch `node` you have to reshim everything in order to work. Sometimes said `reshim` doesn't work, and you need to reinstall anyway. Sometimes both don't work and you need to delete the shims folder and do it all over again.

Comment (jfranciscosousa) [reply bin/setup:43]: `yarn2` should be installed with NPM, but the new version is built on that premise. The new `yarn` basically maintains a separate executable for each version, and it supports multiple yarn versions with the same installation out of the box. But it still bugs out with `asdf` and pretty much every single package is not ready to be used with it.

When `yarn` 2 becomes mandatory, it's `npm` time.
